frontend/public/components/utils/k8s-watch-hook.ts

```diff
@@ -0,0 +1,157 @@
+              : null;
+            ids[r.prop] = { id, dispatch };
+            return ids;
+          }, {})
+        : {},
+    [resources, k8sModels],
+  );
+
+  const dispatch = useDispatch();
+  React.useEffect(() => {
+    Object.keys(reduxIDs)
+      .filter((k) => reduxIDs[k].dispatch)
+      .forEach((k) => dispatch(reduxIDs[k].dispatch));
+    return () => {
+      Object.keys(reduxIDs)
+        .filter((k) => reduxIDs[k].dispatch)
+        .forEach((k) => dispatch(k8sActions.stopK8sWatch(reduxIDs[k].id)));
+    };
+  }, [dispatch, reduxIDs]);
+
+  const resourceK8sSelectorCreator = React.useMemo(
+    () =>
+      createSelectorCreator(
+        //specifying createSelectorCreator<ImmutableMap<string, K8sKind>> throws type error
+        defaultMemoize as <F extends Function>(
+          func: F,
+          equalityCheck?: (
+            a: ImmutableMap<string, K8sKind>,
+            b: ImmutableMap<string, K8sKind>,
+            index: number,
+          ) => boolean,
+        ) => F,
+        (oldK8s, newK8s) =>
+          Object.keys(reduxIDs).every(
+            (k) => oldK8s.get(reduxIDs[k].id) === newK8s.get(reduxIDs[k].id),
+          ),
+      ),
+    [reduxIDs],
+  );
+
+  const resourceK8sSelector = React.useMemo(
+    () =>
+      resourceK8sSelectorCreator(
+        (state: RootState) => state.k8s,
+        (k8s) =>
+          Object.keys(reduxIDs).reduce((r, key) => {
+            const k8sValue = k8s.get(reduxIDs[key].id);
+            return k8sValue ? r.set(key, k8sValue) : r;
+          }, ImmutableMap<string, any>()),
+      ),
+    [reduxIDs, resourceK8sSelectorCreator],
+  );
+
+  const resourceK8s = useSelector<RootState, ImmutableMap<string, any>>(resourceK8sSelector);
+
+  const results = React.useMemo<WatchK8sResult<R>>(() => {
+    if (Object.keys(reduxIDs).length) {
+      return resources.reduce((acc, r) => {
+        if (resourceK8s.has(r.prop)) {
+          let data;
+          if (r.isList) {
+            data = resourceK8s.getIn([r.prop, 'data']);
+            data = data && data.toArray().map((a) => a.toJSON());
+          } else {
+            data = resourceK8s.getIn([r.prop, 'data']);
+            if (data && data.toJSON) {
+              data = data.toJSON();
+            }
+          }
+          const loaded = resourceK8s.getIn([r.prop, 'loaded']);
+          const loadError = resourceK8s.getIn([r.prop, 'loadError']);
+          acc[r.prop] = { data, loaded, loadError };
+        } else {
+          acc[r.prop] = { data: r.isList ? [] : {}, loaded: true, loadError: new NoModelError() };
+        }
+        return acc;
+      }, {});
+    }
+    return resources.reduce((acc, r) => {
+      acc[r.prop] = { data: r.isList ? [] : {}, loaded: false, loadError: undefined };
+      return acc;
+    }, {});
+  }, [resourceK8s, resources, reduxIDs]);
+
+  return results;
+};
+
+type WatchK8sResult<R extends K8sResourceKind> = {
+  [k: string]: { data: R | R[]; loaded: boolean; loadError: any };
```

Comment:
The `R | R[]` part will complicate consuming code, because then we have to coerce `data` to an array using e.g. `_.castArray`.

Why not just `R[]` and treat the single-result case as one-element array?

Reply:
Changed the type to just `R`. Caller can specify whether the result is array or not `useK8sWatch<Foo | Bar[] | ... >` etc

Reply:
OK, sounds good.